### DISCUSSION TITLE
Running YugabyteDB in a docker container

### DISCUSSION THREAD

**Discussion by LexLuthr** [filecoin-project/curio#112] [Show and tell]
## Install Docker
Before installing YugabyteDB, ensure that you have the Docker runtime installed on your localhost. To download and install Docker, select one of the following environments:

 [Docker for Mac](https://store.docker.com/editions/community/docker-ce-desktop-mac)

 [Docker for CentOS](https://store.docker.com/editions/community/docker-ce-server-centos)

 [Docker for Ubuntu](https://store.docker.com/editions/community/docker-ce-server-ubuntu)

 [Docker for Debian](https://store.docker.com/editions/community/docker-ce-server-debian)

 [Docker for Windows](https://store.docker.com/editions/community/docker-ce-desktop-windows)

## Pull the YugabyteDB Docker Image
Open a terminal and pull the YugabyteDB Docker image:

```
docker pull yugabytedb/yugabyte:2024.1.0.0-b129
```
## Create a Directory for Yugabyte Home
Create a local directory to be used as the Yugabyte home directory inside the container:

```
mkdir -p /path/to/yugabyte-home
```

Replace `/path/to/yugabyte-home` with your desired path. This path will be used to store all YugabyteDB related data.

## Run the YugabyteDB Container
Run the following command to start the YugabyteDB container with the specified ports bound and the local directory mounted as a volume:

```
docker run -d --name yugabytedb \
    -p7000:7000 \
    -p7100:7100 \
    -p9000:9000 \
    -p9100:9100 \
    -p5433:5433 \
    -p15433:15433\
    -v /path/to/yugabyte-home:/home/yugabyte/yb_data \
    yugabytedb/yugabyte:2024.1.0.0-b129 bin/yugabyted start --base_dir=/home/yugabyte/yb_data --background=false
```
 ### Explanation of the Command:

|Command Part | Explanation |
|--|--|
| docker run -d | Run the container in detached mode otherwise container lifetime will tied to your terminal session |
| --name yugabytedb | Name the container yugabytedb |
| -p <host_port>:<container_port> | Map the container ports to the host ports |
| 7000 | Master Admin UI |
| 7100 | Master RPC |
| 9000 | TServer Admin UI |
| 9100 | TServer RPC | 
| 9042 | YCQL port |
| 5433 | YSQL port |
| 15433 | YugabyteDB UI |
| -v /path/to/yugabyte-home:/home/yugabyte/yb_data | Mount the local directory to /home/yugabyte/yb_data inside the container |
| yugabytedb/yugabyte:2.5.0.0-b2 bin/yugabyted start --base_dir=/home/yugabyte | Start the YugabyteDB server with the base directory set to the mounted volume |

## Verify the Deployment
Check the running container with:

```
docker ps
```

## Access YugabyteDB
Access the YugabyteDB web interfaces using the http://localhost:15433
Connect to the database using YCQL or YSQL clients on ports 9042 and 5433, respectively.

You've successfully deployed a YugabyteDB container on Docker with the specified ports bound and a local filesystem mounted as the Yugabyte home directory.

**Comment by gsantopaolo** [comment]
Hi @LexLuthr  did you ever tested to run yuga on docker running on a mac? were you able to persist data? if yes, can you share any compose snippet?

**Reply by LexLuthr** [reply]
Yes. Above steps describe exactly that. If you could tell me a little more about your use case, I would be able to provide you with exact info you need. 

https://github.com/filecoin-project/curio/blob/main/docker/docker-compose.yaml has a working example. I use it on MacOS.

**Reply by gsantopaolo** [reply]
Hi there, thank you a lot for your sample; it made my day! 
Here is [my final solution](https://github.com/gsantopaolo/YugabyteDB_sample/tree/main/deployment)
Just a note: bind mounts (like ` ./data/yugabyte:/home/yugabyte/yb_data`) are served from the Mac host via VirtioFS (or, if you hadn’t enabled VirtioFS, via gRPC FUSE). Neither of these is ext4, which is the only file system compatible with yougabylte.
The only way I found to serve a file system ext4 on a Mac was via named volumes. When you use a named volume, Docker stores its data inside the Linux VM’s disk image, which is formatted as an ext4 filesystem.

So I did something like:

```
volumes:
  yugabytedb-data:
```
  
And then in the service definition:
```
olumes:
      - yugabytedb-data:/home/yugabyte/yb_data
```


This way, data on Yuga is persisted on the disk. But, there is a but... if you stop and restart your container, when the container starts up, it wipes all the data.
To avoid that behaviour, I added:


```
    init: true
    environment:
      - TINI_SUBREAPER=true
```

This prevents the data from being wiped out when the container restarts.

Again, thanks a lot!
